spokestack/asr/google/speech_recognizer.py

```diff
@@ -0,0 +1,119 @@
+from google.cloud import speech  # type: ignore
+from google.oauth2 import service_account  # type: ignore
+
+from spokestack.context import SpeechContext
+
+
+_LOG = logging.getLogger(__name__)
+
+
+class GoogleSpeechRecognizer:
+    """Transforms speech into text using Google's ASR.
+
+    Args:
+        language (str): language of input audio
+        language (str): language of input audio
+        credentials (optional): google API credentials or path to credentials.
+                                if set to None these will be pulled from the
+                                environment variable: GOOGLE_APPLICATION_CREDENTIALS
+        sample_rate (int): sample rate of the input audio (Hz)
+        **kwargs (optional): additional keyword arguments
+    """
+
+    def __init__(
+        self, language, credentials=None, sample_rate=16000, **kwargs,
+    ) -> None:
+        if credentials:
+            if isinstance(credentials, str):
+                credentials = service_account.Credentials.from_service_account_file(
+                    credentials
+                )
+            else:
+                credentials = service_account.Credentials.from_service_account_info(
+                    credentials
+                )
+
+        self._client = speech.SpeechClient(credentials=credentials)
+        self._config = speech.StreamingRecognitionConfig(
+            config=speech.RecognitionConfig(
+                encoding=speech.RecognitionConfig.AudioEncoding.LINEAR16,
+                sample_rate_hertz=sample_rate,
+                language_code=language,
+                enable_automatic_punctuation=True,
+            ),
+            interim_results=True,
+        )
+        self._sample_rate = sample_rate
+        self._language = language
+        self._queue: Queue = Queue()
+        self._thread: Any = None
+
+    def __call__(self, context: SpeechContext, frame: np.ndarray) -> None:
+        """Main entry point.
+
+        Args:
+            context (SpeechContext): current state of the speech pipeline
+            frame (np.ndarray): numpy array of PCM-16 audio.
+
+        Returns: None
+
+        """
+        if self._thread is None and context.is_active:
+            self._begin(context)
+        if self._thread is not None and not context.is_active:
+            self._commit()
+        if context.is_active:
+            self._send(frame)
+
+    def _begin(self, context) -> None:
+        self._thread = Thread(target=self._receive, args=(context,),)
+        self._thread.start()
+
+    def _receive(self, context):
+
+        for response in self._client.streaming_recognize(self._config, self._drain()):
+            for result in response.results[:1]:
+                for alternative in result.alternatives[:1]:
+                    context.transcript = alternative.transcript
+                    context.confidence = alternative.confidence
+
+                if result.is_final:
+                    if context.transcript:
```

Comment:
the spokestack recognizer raises the timeout event on the final transcript if it is empty, this recognizer should be consistent with that